src/java/org/apache/cassandra/index/sai/disk/format/IndexDescriptor.java

```diff
@@ -55,267 +53,360 @@
-                                                                  sstable.descriptor,
-                                                                  sstable.metadata().partitioner,
-                                                                  sstable.metadata().comparator);
-
-            if (version.onDiskFormat().isPerSSTableBuildComplete(indexDescriptor))
-            {
-                indexDescriptor.registerPerSSTableComponents();
-                return indexDescriptor;
-            }
+            if (componentExistsOnDisk(version, sstable.descriptor, IndexComponent.GROUP_COMPLETION_MARKER, null))
+                return new IndexDescriptor(version,
+                                           sstable.descriptor,
+                                           sstable.metadata().partitioner,
+                                           sstable.metadata().comparator);
         }
+        // we always want a non-null IndexDescriptor, even if it's empty
         return new IndexDescriptor(Version.LATEST,
                                    sstable.descriptor,
                                    sstable.metadata().partitioner,
                                    sstable.metadata().comparator);
     }
 
-    public boolean hasComponent(IndexComponent indexComponent)
+    public boolean hasComponent(IndexComponent component)
     {
         registerPerSSTableComponents();
-        return perSSTableComponents.contains(indexComponent);
+        return components.get(null).contains(component);
+    }
+
+    public boolean hasComponent(IndexComponent component, IndexContext context)
+    {
+        registerPerIndexComponents(context);
+        var components = this.components.get(context);
+        return components != null && components.contains(component);
     }
 
-    public boolean hasComponent(IndexComponent indexComponent, IndexContext indexContext)
+    public String componentFileName(IndexComponent component)
     {
-        registerPerIndexComponents(indexContext);
-        return perIndexComponents.containsKey(indexContext.getIndexName()) &&
-               perIndexComponents.get(indexContext.getIndexName()).contains(indexComponent);
+        return versions.get(null).fileNameFormatter().format(component, null);
     }
 
-    public String componentName(IndexComponent indexComponent)
+    public String componentFileName(IndexComponent component, IndexContext context)
     {
-        return version.fileNameFormatter().format(indexComponent, null);
+        return getVersion(context).fileNameFormatter().format(component, context);
     }
 
-    public String componentName(IndexComponent indexComponent, IndexContext indexContext)
+    public Version getVersion()
     {
-        return version.fileNameFormatter().format(indexComponent, indexContext);
+        return getVersion(null);
+    }
+
+    public Version getVersion(IndexContext context)
+    {
+        return versions.computeIfAbsent(context, __ ->
```

Comment:
Initially, I was confused about using the `IndexContext` as a key here due to collisions across sstables, but then I realized we have `IndexDescriptor` objects per sstable.